### PR TITLE
add nixos module

### PR DESCRIPTION
**Pull request by brainrake**
Closes # .

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - `Ctl.Examples.PlutusV2.OneShotMinting` example, demonstrating the use of `applyArgs` with a PlutusV2 script ([#1038](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1038))
 - New `Contract.Test.Utils` assertions and checks: `assertOutputHasRefScript`, `checkOutputHasRefScript`, `checkTxHasMetadata` ([#1044](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1044))
 - `Parallel` instance to `Contract` monad. Parallel capabilities are in the associated `ParContract` datatype ([#1037](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1037))
+- NixOS module
```
Comment (ngua) [inline CHANGELOG.md:41]: Could you add a bit more to the description please, maybe just mention that it's for the runtime? Users might not immediately understand this if they're not familiar with CTL and its runtime

File: flake.nix [hunk 2]
```diff
@@ -18,6 +18,13 @@
 
     plutip.url = "github:mlabs-haskell/plutip/8364c43ac6bc9ea140412af9a23c691adf67a18b";
     ogmios-datum-cache.url = "github:mlabs-haskell/ogmios-datum-cache/880a69a03fbfd06a4990ba8873f06907d4cd16a7";
+
+    # ogmios and ogmios-datum-cache nixos modules (remove and replace with the above after merging and updating)
+    ogmios-nixos.url = "github:mlabs-haskell/ogmios/marton/nixos-module";
+    ogmios-datum-cache-nixos.url = "github:mlabs-haskell/ogmios-datum-cache/marton/nixos-module";
+
+    cardano-node.url = "github:input-output-hk/cardano-node";
```
Comment (ngua) [inline flake.nix:26]: `cardano-node` should be pinned to an explicit revision or ref. Currently we're on `v1.35.3`

File: flake.nix [hunk 3]
```diff
@@ -366,6 +378,24 @@
         };
       };
 
+      nixosModules.ctl-server = { pkgs, lib, ... }: {
+        imports = [ ./nix/ctl-server-nixos-module.nix ];
+        services.ctl-server.package = lib.mkDefault self.packages.${pkgs.system}."ctl-server:exe:ctl-server";
```
Comment (ngua) [inline flake.nix:383]: This line is a bit long, can you break it up please?

File: flake.nix [hunk 4]
```diff
@@ -366,6 +378,24 @@
...
+      };
+
+      nixosConfigurations.test = nixpkgs.lib.nixosSystem {
+        system = "x86_64-linux";
+        modules = [
+          inputs.cardano-node.nixosModules.cardano-node
+          inputs.ogmios-nixos.nixosModules.ogmios
+          { services.ogmios.package = inputs.ogmios.packages.x86_64-linux."ogmios:exe:ogmios"; }
```
Comment (ngua) [inline flake.nix:391]: Also quite long (applies to lines 392 and 393 as well)

File: nix/ctl-server-nixos-module.nix [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      isSystemUser = true;
+      group = cfg.group;
+      extraGroups = [ "cardano-node" ];
+    };
+    users.groups.ctl-server = mkIf (cfg.group == "ctl-server") { };
+
+    systemd.services.ctl-server = {
+      enable = true;
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" "postgresql.service" "cardano-node.service" "ogmios.service" ];
```
Comment (ngua) [inline nix/ctl-server-nixos-module.nix:50]: Also a long line

File: nix/ctl-server-nixos-module.nix [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        ProtectControlGroups = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        LockPersonality = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        RemoveIPC = true;
+        PrivateMounts = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [ "~@cpu-emulation @debug @keyring @mount @obsolete @privileged @setuid @resources" ];
```
Comment (ngua) [inline nix/ctl-server-nixos-module.nix:84]: The single element could be on its own line to make it shorter

**Review by ngua — CHANGES_REQUESTED** [review]
I know this is still a draft but I thought a preliminary review would be a good idea

We try to keep lines 80 characters wide or shorter, including in Nix

The module itself looks pretty good, thanks for taking care of this! This will be quite helpful I think

**Review by ngua — COMMENTED** [review]
Could you also add something to `doc/runtime.md` explaining how to use this and when it might be useful for users?

File: CHANGELOG.md (hunk 1 above)
Comment (brainrake) [reply CHANGELOG.md:41]: updated

**Comment by brainrake** [comment]
Thanks for the review! Updated.

**Comment by aciceri** [comment]
Closing this in favor of #1194

**Pull request closed without merging**